### PR TITLE
Problem: no integration test for multiple attestation processing (gravity bridge)

### PR DESCRIPTION
**Pull request by thomas-nguy**
👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

close https://github.com/crypto-org-chain/cronos/issues/726

# PR Checklist:

- [ ] Have you read the [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md)?
- [ ] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [ ] Have you rebased your work on top of the latest master? 
- [ ] Have you checked your code compiles? (`make`)
- [ ] Have you included tests for any non-trivial functionality?
- [ ] Have you checked your code passes the unit tests? (`make test`)
- [ ] Have you checked your code formatting is correct? (`go fmt`)
- [ ] Have you checked your basic code style is fine? (`golangci-lint run`)
- [ ] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [ ] If your changes affect the client infrastructure, have you run the integration test?
- [ ] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [ ] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-org-chain/chain-main/blob/master/CHANGELOG.md)?
- [ ] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-org-chain/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

File: integration_tests/utils.py [hunk 1]
```diff
@@ -506,3 +506,40 @@ def send_txs(w3, cli, to, keys, params):
...
+            token_contract.address, HexBytes(recipient), amount
+        ).buildTransaction({"from": acct_address})
+        signed = sign_transaction(w3, tx, key_from)
+        raw_transactions.append(signed.rawTransaction)
+
+    # wait for new block
+    w3_wait_for_new_blocks(w3, 1)
+
+    # send transactions
+    with ThreadPoolExecutor(len(raw_transactions)) as exec:
```
Comment (mmsqe) [inline integration_tests/utils.py:539]: maybe can do 2 batches, approve and send

File: integration_tests/test_gravity.py [hunk 2]
```diff
@@ -303,7 +305,61 @@ def check():
...
+            check the all attestation are processed at once by comparing
+            with previous block balance
+            """
+            nonlocal previous
+            current = cli.balance(eth_to_bech32(recipient), denom=denom)
+            check = current == previous + (10 * len(ACCOUNTS))
+            previous = current
+            return check
+
+        # we should interval value lower than block internal to check
```
Comment (adu-crypto) [inline integration_tests/test_gravity.py:357]:
```suggestion
        # we should interval value lower than block interval to check
```

**Review by adu-crypto — APPROVED** [review]
lgtm

File: integration_tests/utils.py (hunk 1 above)
Comment (thomas-nguy) [reply integration_tests/utils.py:539]: could do, but its less critical for the approval which do not requires being included in the same block (or block+1)

File: integration_tests/utils.py [hunk 3]
```diff
@@ -506,3 +506,40 @@ def send_txs(w3, cli, to, keys, params):
...
+def multiple_send_to_cosmos(gravity_contract, token_contract, recipient, amount, keys):
+    # use different sender accounts to be able be send concurrently
+    w3 = token_contract.web3
+    raw_transactions = []
+    for key_from in keys:
+        acct = Account.from_key(key_from)
+        acct_address = HexBytes(acct.address)
+        # approve first
+        txreceipt = send_transaction(
+            token_contract.web3,
```
Comment (mmsqe) [inline integration_tests/utils.py:520]:
```suggestion
            w3,
```

File: integration_tests/utils.py (hunk 1 above)
Comment (thomas-nguy) [reply integration_tests/utils.py]: sure 👌

File: integration_tests/test_gravity.py [hunk 4]
```diff
@@ -303,7 +305,61 @@ def check():
...
+            with previous block balance
+            """
+            nonlocal previous
+            current = cli.balance(eth_to_bech32(recipient), denom=denom)
+            check = current == previous + (10 * len(ACCOUNTS))
+            previous = current
+            return check
+
+        # we choose an interval value lower than block interval to check
+        # that all attestation are processing within the same block
```
Comment (yihuang) [inline integration_tests/test_gravity.py:358]:
```suggestion
        # that all attestation are processed within the same block
```

File: integration_tests/test_gravity.py [hunk 5]
```diff
@@ -303,7 +305,61 @@ def check():
...
+            gravity.contract, erc20, recipient, amount, KEYS.values()
+        )
+
+        def check_gravity_balance():
+            """
+            check the all attestation are processed at once by comparing
+            with previous block balance
+            """
+            nonlocal previous
+            current = cli.balance(eth_to_bech32(recipient), denom=denom)
```
Comment (yihuang) [inline integration_tests/test_gravity.py:352]: Alternatively, you can pass the height parameter explicitly

Comment (thomas-nguy) [reply integration_tests/test_gravity.py]: sure

File: integration_tests/test_gravity.py [hunk 6]
```diff
@@ -343,20 +343,23 @@ def test_multiple_attestation_processing(gravity):
...
+        current_height = cli.block_height()
         def check_gravity_balance():
             """
             check the all attestation are processed at once by comparing
             with previous block balance
             """
             nonlocal previous
-            current = cli.balance(eth_to_bech32(recipient), denom=denom)
+            nonlocal current_height
+            current = cli.balance(eth_to_bech32(recipient), denom=denom, height=current_height)
```
Comment (yihuang) [inline integration_tests/test_gravity.py:354]: what would be the result if height is larger than latest one, we should check against with current height?

Comment (thomas-nguy) [reply integration_tests/test_gravity.py]: we are incrementing the height by one at the end of the loop, so we are checking each block from first current_height

Comment (thomas-nguy) [reply integration_tests/test_gravity.py]: to be more safe, I will get the height before sending the tx

**Review by mmsqe — APPROVED** [review]

**Pull request merged**